### PR TITLE
Refactor hash join with cuCollections multimap

### PR DESCRIPTION
**Pull request by PointKernel**
This PR refactors the existing hash join implementation by using cuCollections `static_multimap`. Related functions now invoke `cuco::static_multimap` host-bulk APIs thus join-specific kernels are no longer needed. Compared to the current cudf multimap, `cuco::static_multimap` applied a list of optimizations to improve overall performance:
- Using CUDA Cooperative Groups & parallel retrieval to leverage multiple threads for a single hash map operation
- Double hashing probe method instead of linear probing to reduce the chance of collisions
- Relaxed atomic operations instead of expensive sequential memory order atomic
- Using vector loads for "packable" pairs (no larger than 4B/4B) to coarsen the workload granularity for each thread

`inner_join` benchmarks are used to evaluate new hash join performance: 
- multiplicity = 1
- matching rate = 0.3
- occupancy = 0.5
- double hashing
- CG size = 2
- 75% nulls if nulls are present:
<details><summary>inner_join_32bit</summary>
<p>

## [0] Tesla V100-SXM2-32GB
|  Key Type  |  Payload Type  |  Nullable  |  Build Table Size  |  Probe Table Size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |       Diff |   %Diff |  Status  |
|------------|----------------|------------|--------------------|--------------------|------------|-------------|------------|-------------|------------|---------|----------|
|    I32     |      I32       |     0      |       100000       |       100000       | 135.391 us |      15.35% | 268.328 us |      18.26% | 132.937 us |  98.19% |   PASS   |
|    I32     |      I32       |     0      |       100000       |       400000       | 248.937 us |       5.55% | 364.678 us |      10.10% | 115.740 us |  46.49% |   PASS   |
|    I32     |      I32       |     0      |      10000000      |      10000000      |  12.635 ms |       0.10% |  13.580 ms |       0.13% | 945.092 us |   7.48% |   PASS   |
|    I32     |      I32       |     0      |      10000000      |      40000000      |  37.934 ms |       0.05% |  43.083 ms |       0.08% |   5.148 ms |  13.57% |   PASS   |
|    I32     |      I32       |     0      |      10000000      |     100000000      |  88.486 ms |       0.20% | 108.750 ms |       0.05% |  20.264 ms |  22.90% |   PASS   |
|    I32     |      I32       |     0      |      80000000      |     100000000      | 124.207 ms |       0.04% | 130.181 ms |       0.04% |   5.974 ms |   4.81% |   PASS   |
|    I32     |      I32       |     0      |     100000000      |     100000000      | 134.720 ms |       0.12% | 139.181 ms |       0.03% |   4.462 ms |   3.31% |   PASS   |
|    I32     |      I32       |     0      |      10000000      |     240000000      | 218.672 ms |       0.04% | 295.274 ms |       0.03% |  76.602 ms |  35.03% |   PASS   |
|    I32     |      I32       |     0      |      80000000      |     240000000      | 242.677 ms |       0.03% | 266.421 ms |       0.04% |  23.744 ms |   9.78% |   PASS   |
|    I32     |      I32       |     0      |     100000000      |     240000000      | 253.245 ms |       0.04% | 273.783 ms |       0.04% |  20.538 ms |   8.11% |   PASS   |
</p>
</details>

<details><summary>inner_join_64bit</summary>
<p>

## [0] Tesla V100-SXM2-32GB

|  Key Type  |  Payload Type  |  Nullable  |  Build Table Size  |  Probe Table Size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |      Diff |   %Diff |  Status  |
|------------|----------------|------------|--------------------|--------------------|------------|-------------|------------|-------------|-----------|---------|----------|
|    I64     |      I64       |     0      |      40000000      |      50000000      |  62.728 ms |       0.04% |  65.638 ms |       0.07% |  2.910 ms |   4.64% |   PASS   |
|    I64     |      I64       |     0      |      50000000      |      50000000      |  68.142 ms |       0.05% |  70.164 ms |       0.05% |  2.022 ms |   2.97% |   PASS   |
|    I64     |      I64       |     0      |      40000000      |     120000000      | 122.424 ms |       0.04% | 134.553 ms |       0.05% | 12.129 ms |   9.91% |   PASS   |
|    I64     |      I64       |     0      |      50000000      |     120000000      | 127.979 ms |       0.04% | 138.379 ms |       0.04% | 10.400 ms |   8.13% |   PASS   |
</p>
</details>

<details><summary>inner_join_32bit_nulls</summary>
<p>

## [0] Tesla V100-SXM2-32GB

|  Key Type  |  Payload Type  |  Nullable  |  Build Table Size  |  Probe Table Size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|------------|----------------|------------|--------------------|--------------------|------------|-------------|------------|-------------|---------------|---------|----------|
|    I32     |      I32       |     1      |       100000       |       100000       | 118.400 us |      10.77% | 237.164 us |      10.76% |    118.764 us | 100.31% |   PASS   |
|    I32     |      I32       |     1      |       100000       |       400000       | 167.385 us |       7.46% | 275.823 us |       6.00% |    108.438 us |  64.78% |   PASS   |
|    I32     |      I32       |     1      |      10000000      |      10000000      |   3.699 ms |       0.28% |   3.053 ms |       0.24% |   -646.276 us | -17.47% |   PASS   |
|    I32     |      I32       |     1      |      10000000      |      40000000      |  10.574 ms |       0.15% |   8.634 ms |       0.08% |  -1939.859 us | -18.35% |   PASS   |
|    I32     |      I32       |     1      |      10000000      |     100000000      |  24.461 ms |       0.11% |  19.744 ms |       0.05% |  -4716.875 us | -19.28% |   PASS   |
|    I32     |      I32       |     1      |      80000000      |     100000000      |  35.013 ms |       0.11% |  27.576 ms |       0.05% |  -7437.653 us | -21.24% |   PASS   |
|    I32     |      I32       |     1      |     100000000      |     100000000      |  37.870 ms |       0.42% |  29.800 ms |       0.06% |  -8069.778 us | -21.31% |   PASS   |
|    I32     |      I32       |     1      |      10000000      |     240000000      |  58.529 ms |       0.19% |  46.316 ms |       0.02% | -12213.277 us | -20.87% |   PASS   |
|    I32     |      I32       |     1      |      80000000      |     240000000      |  67.443 ms |       0.12% |  53.721 ms |       0.49% | -13722.214 us | -20.35% |   PASS   |
|    I32     |      I32       |     1      |     100000000      |     240000000      |  70.809 ms |       0.50% |  55.938 ms |       0.03% | -14870.980 us | -21.00% |   PASS   |
</p>
</details>

<details><summary>inner_join_64bit_nulls</summary>
<p>

## [0] Tesla V100-SXM2-32GB

|  Key Type  |  Payload Type  |  Nullable  |  Build Table Size  |  Probe Table Size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|------------|----------------|------------|--------------------|--------------------|------------|-------------|------------|-------------|--------------|---------|----------|
|    I64     |      I64       |     1      |      40000000      |      50000000      |  17.560 ms |       0.21% |  13.926 ms |       0.09% | -3634.422 us | -20.70% |   PASS   |
|    I64     |      I64       |     1      |      50000000      |      50000000      |  19.070 ms |       0.14% |  15.082 ms |       0.07% | -3987.592 us | -20.91% |   PASS   |
|    I64     |      I64       |     1      |      40000000      |     120000000      |  34.253 ms |       0.13% |  27.170 ms |       0.05% | -7082.957 us | -20.68% |   PASS   |
|    I64     |      I64       |     1      |      50000000      |     120000000      |  35.807 ms |       0.19% |  28.340 ms |       0.03% | -7467.177 us | -20.85% |   PASS   |

</p>
</details>

When nulls are not present, the map has an actual occupancy of 0.5. The new hash join implementation (`Cmp`) outperforms the existing one (`Ref`) by 20% (can be achieved by tuning CG size) to 100%. When nulls are present and treated as unequal, however, the actual occupancy is 0.125 and `Cmp` is always about 20% slower than `Ref`. Using the number of valid rows (https://github.com/rapidsai/cudf/issues/9176) to build hash map can solve this performance issue. Note that the above results show the minimum speedups brought by the new hash join since multiplicity and load factor are relatively low in this case. 

Also, this PR simplifies and improves the current join benchmark by adding multiplicity control. It fixs bugs in join cpp tests and pytests where outputs were not sorted before compare.

### CODE REVIEW

**Comment by jrhemstad** [comment]
More lines removed than added. That's what I like to see!

File: cpp/src/join/hash_join.cu [hunk 1]
```diff
@@ -32,6 +33,23 @@
 namespace cudf {
 namespace detail {
 
+namespace {
+
+class build_predicate {
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:38]: I'd rename this to be something more descriptive like `row_contains_null`

File: cpp/src/join/hash_join.cu [hunk 2]
```diff
@@ -64,28 +78,25 @@ std::unique_ptr<multimap_type, std::function<void(multimap_type*)>> build_join_h
...
   std::size_t const hash_table_size = compute_hash_table_size(build_table_num_rows);
 
-  auto hash_table = multimap_type::create(hash_table_size,
-                                          stream,
-                                          true,
-                                          multimap_type::hasher(),
-                                          multimap_type::key_equal(),
-                                          multimap_type::allocator_type());
+  auto hash_table = std::make_unique<multimap_type>(
+    hash_table_size, std::numeric_limits<hash_value_type>::max(), JoinNoneValue);
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:82]: I don't think the `unique_ptr` is necessary here. You can just construct the `static_map` as normal.

File: cpp/src/join/hash_join.cu [hunk 3]
```diff
@@ -64,28 +78,25 @@ std::unique_ptr<multimap_type, std::function<void(multimap_type*)>> build_join_h
...
+
+  row_hash hash_build{*build_device_table};
+  auto const empty_key_sentinel = hash_table->get_empty_key_sentinel();
+  make_pair_function pair_func{hash_build, empty_key_sentinel};
+
+  thrust::counting_iterator<size_type> first(0);
+  thrust::transform_iterator<make_pair_function,
+                             thrust::counting_iterator<size_type>,
+                             cudf::detail::pair_type>
+    iter(first, pair_func);
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:97]: Use `make_transform_iterator`.

File: cpp/cmake/thirdparty/CUDF_GetCUCO.cmake [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  if(CUCO_INCLUDE)
+    set(CUCO_INCLUDE_DIR "${CUCO_INCLUDE}" PARENT_SCOPE)
+    return()
+  endif()
+  if(CUCO_INCLUDE_DIR)
+    set(CUCO_INCLUDE_DIR ${CUCO_INCLUDE_DIR} PARENT_SCOPE)
+    return()
+  endif()
+  CPMFindPackage(NAME cuco
+    GITHUB_REPOSITORY PointKernel/cuCollections
```
Comment (devavret) [inline cpp/cmake/thirdparty/CUDF_GetCUCO.cmake:27]: Should this be NVIDIA/cuCollections?

Comment (PointKernel) [reply cpp/cmake/thirdparty/CUDF_GetCUCO.cmake:27]: The multimap branch hasn't been merged into the NVIDIA repo yet. Will update this once all reviews/issues in the current PR are addressed.

File: cpp/src/join/hash_join.cu [hunk 5]
```diff
@@ -43,51 +61,36 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> get_empty_joined_table
...
+ * @param hash_table Build hash table.
  * @param compare_nulls Controls whether null join-key values should match or not.
  * @param stream CUDA stream used for device memory operations and kernel launches.
  *
- * @return Built hash table.
  */
-std::unique_ptr<multimap_type, std::function<void(multimap_type*)>> build_join_hash_table(
-  cudf::table_view const& build, null_equality compare_nulls, rmm::cuda_stream_view stream)
+void build_join_hash_table(cudf::table_view const& build,
+                           multimap_type& hash_table,
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:71]: Can't this return the hash_table?

File: cpp/src/join/hash_join.cu [hunk 6]
```diff
@@ -43,51 +61,36 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> get_empty_joined_table
...
+
+  row_hash hash_build{*build_device_table};
+  auto const empty_key_sentinel = hash_table.get_empty_key_sentinel();
+  make_pair_function pair_func{hash_build, empty_key_sentinel};
+
+  thrust::counting_iterator<size_type> first(0);
+  auto iter = thrust::make_transform_iterator(first, pair_func);
+
+  size_type const build_table_num_rows{build_device_table->num_rows()};
+  hash_table.insert_if(iter, iter + build_table_num_rows, pred, stream.value());
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:93]: If using the `insert_if` when the row_bitmask is null proves to be a performance limiter, then we should look at having an `if/else` for using `insert` vs `insert_if`.

File: cpp/src/join/hash_join.cuh [hunk 7]
```diff
@@ -39,6 +38,38 @@
...
+ public:
+  make_pair_function(row_hash const& hash, hash_value_type const empty_key_sentinel)
+    : _hash{hash}, _empty_key_sentinel{empty_key_sentinel}
+  {
+  }
+
+  __device__ __inline__ cudf::detail::pair_type operator()(size_type i) const noexcept
+  {
+    auto row_hash_value = remap_sentinel_hash(_hash(i), _empty_key_sentinel);
+    return cuco::make_pair<hash_value_type, size_type>(std::move(row_hash_value), std::move(i));
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cuh:65]: We need to resolve this `std::move` thing. That should not be necessary.

Comment (PointKernel) [reply cpp/src/join/hash_join.cuh:65]: I've worked around https://github.com/NVIDIA/cuCollections/pull/87 a bit but didn't find a solution.

**Review by shwina — APPROVED** [review]
Approving Python changes

File: cpp/src/join/hash_join.cu [hunk 8]
```diff
@@ -43,51 +61,38 @@ std::pair<std::unique_ptr<table>, std::unique_ptr<table>> get_empty_joined_table
...
-  // Check error code from the kernel
-  if (failure.value(stream) == 1) { CUDF_FAIL("Hash Table insert failure."); }
-
-  return hash_table;
+  size_type const build_table_num_rows{build_device_table->num_rows()};
+  if (compare_nulls == null_equality::EQUAL) {
+    hash_table.insert(iter, iter + build_table_num_rows, stream.value());
+  } else {
+    auto const row_bitmask = cudf::detail::bitmask_and(build, stream);
+    row_contains_null pred{static_cast<bitmask_type const*>(row_bitmask.data())};
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:92]: There's an opportunity for optimization here where we don't need to build the `row_bitmask` when none of the columns in the build table contain nulls. 
```suggestion
  if ((compare_nulls == null_equality::EQUAL) or (not nullable(build))) {
    hash_table.insert(iter, iter + build_table_num_rows, stream.value());
  } else {
    auto const row_bitmask = cudf::detail::bitmask_and(build, stream);
    row_contains_null pred{static_cast<bitmask_type const*>(row_bitmask.data())};
```

File: cpp/cmake/thirdparty/get_cucollections.cmake [hunk 9]
```diff
@@ -22,11 +22,13 @@ function(find_and_configure_cucollections)
 
     # Find or install cuCollections
     CPMFindPackage(NAME   cuco
-        GITHUB_REPOSITORY PointKernel/cuCollections
-        GIT_TAG           static-multi-map
-        OPTIONS           "BUILD_TESTS OFF"
-                          "BUILD_BENCHMARKS OFF"
-                          "BUILD_EXAMPLES OFF"
+        GLOBAL_TARGETS cuco::cuco
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/get_cucollections.cmake:25]: GLOBAL_TARGETS and CPM_ARGS are features of `rapids-cmake` and the `rapids_cpm_find` function. 

With https://github.com/rapidsai/cudf/pull/9030 being merged you should change `CPMFindPackage` to `rapids_cpm_find`.

Comment (PointKernel) [reply cpp/cmake/thirdparty/get_cucollections.cmake:25]: Done.

File: cpp/cmake/thirdparty/get_cucollections.cmake [hunk 10]
```diff
@@ -14,18 +14,20 @@
 # limitations under the License.
 #=============================================================================
 
+# cuCollections doesn't have a version
+
 function(find_and_configure_cucollections)
 
     if(TARGET cuco::cuco)
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/get_cucollections.cmake:21]: Now with [rapids-cmake  21.10](https://docs.rapids.ai/api/rapids-cmake/nightly/command/rapids_cpm_find.html) the `if(TARGET cuco::cuco)` check is not needed. `rapids_cpm_find` will do it internally.

**Review by robertmaynard — APPROVED** [review]
CMake additions LGTM

File: cpp/src/join/hash_join.cu [hunk 11]
```diff
@@ -39,9 +39,9 @@ class row_contains_null {
  public:
   row_contains_null(bitmask_type const* row_bitmask) : _row_bitmask{row_bitmask} {}
 
-  __device__ __inline__ bool operator()(const pair_type& pair) const noexcept
+  __device__ __inline__ bool operator()(const size_type& i) const noexcept
   {
-    return _row_bitmask == nullptr or cudf::bit_is_set(_row_bitmask, pair.second);
+    return _row_bitmask == nullptr or cudf::bit_is_set(_row_bitmask, i);
```
Comment (jrhemstad) [inline cpp/src/join/hash_join.cu:44]: `_row_bitmask == nullptr` can never be true now, right? We'll only ever call `insert_if` when we've materialized a row bitmask.

Comment (PointKernel) [reply cpp/src/join/hash_join.cu:44]: Yes, you are right.

**Comment by PointKernel** [comment]
rerun tests

File: cpp/src/join/hash_join.cu (hunk 5 above)
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:71]: Currently infeasible due to a compiler bug (https://godbolt.org/z/7fqsK4hPr) in `cuda-11.0` and ` cuda-11.4`

File: cpp/src/join/hash_join.cu [hunk 12]
```diff
@@ -449,19 +450,19 @@ hash_join::hash_join_impl::probe_join_indices(cudf::table_view const& probe,
...
+  if (_is_empty and JoinKind != cudf::detail::join_kind::INNER_JOIN) {
     return get_trivial_left_join_indices(probe, stream, mr);
   }
 
-  CUDF_EXPECTS(_hash_table, "Hash table of hash join is null.");
+  CUDF_EXPECTS(!_is_empty, "Hash table of hash join is null.");
 
-  auto build_table = cudf::table_device_view::create(_build, stream);
-  auto probe_table = cudf::table_device_view::create(probe, stream);
+  auto build_table_ptr = cudf::table_device_view::create(_build, stream);
```
Comment (devavret) [inline cpp/src/join/hash_join.cu:459]: nit: no need for `_ptr`

**Review by devavret — CHANGES_REQUESTED** [review]

File: cpp/src/join/hash_join.cu (hunk 12 above)
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:459]: Using `_ptr` suffix is to be consistent with the rest of the code. e.g. https://github.com/rapidsai/cudf/blob/branch-21.10/cpp/src/join/hash_join.cu#L358-L359. And IMO, using `_ptr` is more intuitive. What do you think?

File: cpp/src/join/hash_join.cu [hunk 13]
```diff
@@ -285,7 +282,11 @@ hash_join::hash_join_impl::~hash_join_impl() = default;
 hash_join::hash_join_impl::hash_join_impl(cudf::table_view const& build,
                                           null_equality compare_nulls,
                                           rmm::cuda_stream_view stream)
-  : _hash_table(nullptr)
+  : _is_empty{build.num_rows() == 0},
+    _hash_table{compute_hash_table_size(build.num_rows()),
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:286]: The hash table construction should defer. The idea was to have `build_join_hash_table` return the hash table but there is a compiler bug in `nvcc 11.0` and `nvcc 11.4` stopping us from doing so: https://godbolt.org/z/4Y4e474bM.

File: cpp/src/join/hash_join.cu [hunk 14]
```diff
@@ -32,6 +33,23 @@
 namespace cudf {
 namespace detail {
 
+namespace {
+
+class row_contains_null {
+ public:
+  row_contains_null(bitmask_type const* row_bitmask) : _row_bitmask{row_bitmask} {}
```
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:40]: Good point. Renamed it to `row_is_valid`.

File: cpp/cmake/thirdparty/get_cucollections.cmake [hunk 15]
```diff
@@ -14,14 +14,16 @@
...
 function(find_and_configure_cucollections)
 
     # Find or install cuCollections
     rapids_cpm_find(cuco 0.0
         GLOBAL_TARGETS cuco::cuco
         CPM_ARGS
-            GITHUB_REPOSITORY NVIDIA/cuCollections
-            GIT_TAG           0d602ae21ea4f38d23ed816aa948453d97b2ee4e
+            GITHUB_REPOSITORY PointKernel/cuCollections
+            GIT_TAG           static-multi-map
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/get_cucollections.cmake:26]: What is the ETA on these changes being upstreamed to cuCollections?

Comment (PointKernel) [reply cpp/cmake/thirdparty/get_cucollections.cmake:26]: @jrhemstad may know the answer.

Comment (jrhemstad) [reply cpp/cmake/thirdparty/get_cucollections.cmake:26]: Soon(TM)

**Comment by PointKernel** [comment]
rerun tests

**Review by devavret — APPROVED** [review]

**Comment by shwina** [comment]
@PointKernel Could you please re-target this PR to 21.12? Thanks.

File: cpp/cmake/thirdparty/get_cucollections.cmake (hunk 15 above)
Comment (jrhemstad) [reply cpp/cmake/thirdparty/get_cucollections.cmake:26]: @PointKernel update this to the git hash for the multimap merge.

**Comment by jrhemstad** [comment]
@gpucibot merge

File: cpp/src/join/hash_join.cu (hunk 5 above)
Comment (PointKernel) [reply cpp/src/join/hash_join.cu:75]: Depends on https://github.com/NVIDIA/libcudacxx/issues/205

**Pull request merged**